scripts/gen_dockerfile_test.py

```diff
@@ -0,0 +1,220 @@
+    ('runtime_config:\n python_version:', False, {
+        'dockerfile_python_version': '',
+    }),
+    ('runtime_config:\n python_version: 2', False, {
+        'dockerfile_python_version': '',
+    }),
+    ('runtime_config:\n python_version: 3', False, {
+        'dockerfile_python_version': '3.5',
+    }),
+    ('runtime_config:\n python_version: 3.4', False, {
+        'dockerfile_python_version': '3.4',
+    }),
+    ('runtime_config:\n python_version: 3.5', False, {
+        'dockerfile_python_version': '3.5',
+    }),
+    # requirements.txt present
+    ('env: flex', True, {
+        'has_requirements_txt': True,
+    }),
+    # entrypoint present
+    ('entrypoint: my entrypoint', False, {
+        'entrypoint': 'exec my entrypoint',
+    }),
+])
+def test_get_app_config_valid(app_yaml, isfile, expected):
+    config_file = 'some_config_file'
+    base_image = 'some_image_name'
+    source_dir = 'some_source_dir'
+    raw_app_config = yaml.safe_load(app_yaml)
+    with unittest.mock.patch.object(os.path, 'isfile', return_value=isfile):
+        actual = gen_dockerfile.get_app_config(
+            raw_app_config, base_image, config_file,
+            source_dir)
+        for key, value in expected.items():
+            assert getattr(actual, key) == value
+
+
+@pytest.mark.parametrize('app_yaml', [
+    # Empty app.yaml
+    '',
+    # Invalid entrypoint
+    'entrypoint: "bad \\n entrypoint"',
+    # Invalid python version
+    'runtime_config:\n python_version: 1',
+    'runtime_config:\n python_version: python2',
+])
+def test_get_app_config_invalid(app_yaml):
+    config_file = 'some_config_file'
+    base_image = 'some_image_name'
+    source_dir = 'some_source_dir'
+    raw_app_config = yaml.safe_load(app_yaml)
+    with pytest.raises(ValueError):
+        gen_dockerfile.get_app_config(
+            raw_app_config, base_image, config_file, source_dir)
+
+
+# Basic AppConfig used below
+_base = gen_dockerfile.AppConfig(
+    base_image='',
+    dockerfile_python_version='',
+    entrypoint='',
+    has_requirements_txt=False
+)
+
+
+@pytest.mark.parametrize('app_config, should_find, test_string', [
+    # Requirements.txt
+    (_base, False, 'ADD requirements.txt'),
+    (_base._replace(has_requirements_txt=True), True,
+     'ADD requirements.txt'),
+    # Entrypoint
+    (_base, False, 'CMD'),
+    (_base._replace(entrypoint='my entrypoint'), True,
+     'CMD my entrypoint'),
+    (_base._replace(entrypoint='exec my entrypoint'), True,
+     'CMD exec my entrypoint'),
+    # Base runtime image
+    (_base._replace(base_image='my_base_runtime_image'), True,
+     'FROM my_base_runtime_image'),
+    # Python version
+    (_base._replace(dockerfile_python_version='_my_version'), True,
+     'python_version=python_my_version'),
+])
+def test_generate_files(app_config, should_find, test_string):
+    result = gen_dockerfile.generate_files(app_config)
+    assert sorted(result.keys()) == sorted(EXPECTED_OUTPUT_FILES)
```

Comment:
No action required: I would just use sets for this, make `EXPECTED_OUTPUT_FILES` a set (therefore constant and immutable) and just do `assert set(results.keys()) == EXPECTED_OUTPUT_FILES`.